contracts/ShapeshiftDAORouter.sol

```diff
@@ -0,0 +1,435 @@
+        address recipient,
+        uint256 amount
+    ) external returns (uint256) {
+        return
+            _withdraw(
+                IERC20(token),
+                msg.sender,
+                recipient,
+                amount,
+                0,
+                MAX_VAULT_ID
+            );
+    }
+
+    /**
+     * @notice Called to redeem the caller's shares from underlying vault(s), with the proceeds distributed to recipient.
+     * @dev The caller must approve this contract to use their vault shares or this call will revert.
+     * @param token Address of the ERC20 token to withdraw from vaults
+     * @param recipient Address to receive the withdrawn tokens
+     * @param amount Maximum number of tokens to withdraw from all vaults; actual withdrawal may be less. If `WITHDRAW_EVERYTHING`, just withdraw everything.
+     * @param firstVaultId First vault id to pull from; 0 to start at the the beginning
+     * @param lastVaultId Last vault id to pull from; `MAX_VAULT_ID` to withdraw from all vaults
+     * @return The number of tokens received by recipient.
+     */
+    function withdraw(
+        address token,
+        address recipient,
+        uint256 amount,
+        uint256 firstVaultId,
+        uint256 lastVaultId
+    ) external returns (uint256) {
+        return
+            _withdraw(
+                IERC20(token),
+                msg.sender,
+                recipient,
+                amount,
+                firstVaultId,
+                lastVaultId
+            );
+    }
+
+    /**
+     * @notice Called to redeem withdrawer's shares from underlying vault(s), with the proceeds distributed to recipient.
+     * @dev Withdrawer must approve this contract to use their vault shares or this call will revert.
+     * @param token Address of the ERC20 token to withdraw from vaults
+     * @param withdrawer Address to pull the vault shares from. SECURITY SENSITIVE.
+     * @param recipient Address to receive the withdrawn tokens
+     * @param amount Maximum number of tokens to withdraw from all vaults; actual withdrawal may be less. If `WITHDRAW_EVERYTHING`, just withdraw everything.
+     * @param firstVaultId First vault id to pull from; 0 to start at the the beginning
+     * @param lastVaultId Last vault id to pull from; `MAX_VAULT_ID` to withdraw from all vaults
+     * @return withdrawn The number of tokens received by recipient.
+     */
+    function _withdraw(
+        IERC20 token,
+        address withdrawer,
+        address recipient,
+        uint256 amount,
+        uint256 firstVaultId,
+        uint256 lastVaultId
+    ) internal returns (uint256 withdrawn) {
+        require(firstVaultId <= lastVaultId);
+
+        uint256 _lastVaultId = lastVaultId;
+        if (_lastVaultId == MAX_VAULT_ID)
+            _lastVaultId = registry.numVaults(address(token)) - 1;
+        for (
+            uint256 i = firstVaultId;
+            withdrawn + 1 < amount && i <= _lastVaultId;
+            i++
```

Comment:
nit: missing `;`

Reply:
IDK if it'll compile with a semicolon there, but it's not part of the grammar -- note that the last part of a [`for-statment`](https://docs.soliditylang.org/en/latest/grammar.html#a4.SolidityParser.forStatement) is an [`expression`](https://docs.soliditylang.org/en/latest/grammar.html#a4.SolidityParser.expression) (no trailing semicolon) rather than an [`expression-statement`](https://docs.soliditylang.org/en/latest/grammar.html#a4.SolidityParser.expressionStatement) (trailing semicolon).

Reply:
ah...haha.  Sorry, glanced over the fact that it was in the for loop.